### PR TITLE
add parallel_attention_blocks

### PR DESCRIPTION
**Pull request by lessw2020**
Summary:
This PR adds Parallel Attention blocks to Torch MultiModal. 
There are 3 main additions:
a - Rotary Embeddings 
b - RMS Norm
c - Parallel_Blocks 

Test plan:
Code has been tested separately in ViT and LLM applications and with rotary unit tests.  
However, a general unit test needs to be added along with rotary unit tests being migrated.

### CODE REVIEW

File: torchmultimodal/modules/layers/normalizations.py [hunk 1]
```diff
@@ -45,3 +45,17 @@ def forward(self, x: Tensor) -> Tensor:
             self.eps,
         )
         return output.type_as(x)
+
+
+class RMSNorm(nn.Module):
```
Comment (rohan-varma) [inline torchmultimodal/modules/layers/normalizations.py:50]: add docstring

File: torchmultimodal/modules/layers/parallel_blocks.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+class ParallelAttentionBlock(nn.Module):
+    """
+    Transformer layer multi-head attention and MLP, in a parallelized fashion rather than sequential,
+    with optional attention masking.
+    Inspired by PaLM:  https://arxiv.org/abs/2204.02311
+
+    * We use SwiGLU for the activation function
+    * SwiGLU will approximate same total num params as traditional MLP with GELU
+    * Cross Attention is not enabled here (but available)
```
Comment (rohan-varma) [inline torchmultimodal/modules/layers/parallel_blocks.py:38]: what does "but available" mean?

File: torchmultimodal/modules/layers/parallel_blocks.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    * MQA and GQA are enabled - modify heads via 'num_heads_group_query_attn'
```
Comment (rohan-varma) [inline torchmultimodal/modules/layers/parallel_blocks.py:39]: this seems for GQA, how would I use MQA?

File: torchmultimodal/modules/layers/parallel_blocks.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        attention_dropout: float = 0.0,
+        use_group_query_attention: bool = True,
+        num_heads_group_query_attention: int = 1,
+        use_in_projection_bias: bool = True,
+        use_out_projection_bias: bool = True,
+        use_weight_init: bool = True,
+        num_layers: int = 1,
+        use_rms_norm: bool = True,
+        use_rotary_embeddings: bool = False,
+        max_expected_seq_len: int = 2048,  # needed only if using rotary
```
Comment (rohan-varma) [inline torchmultimodal/modules/layers/parallel_blocks.py:64]: so if I use rotary embeddings, my sequence length is fixed for pretraining, finetuning, inference? cc @ankitade

**Review by rohan-varma — COMMENTED** [review]
looking forward to the unittests, will do more thorough review after that

File: torchmultimodal/modules/layers/parallel_blocks.py (hunk 2 above)
Comment (lessw2020) [reply torchmultimodal/modules/layers/parallel_blocks.py:38]: ah - we support cross attention in the main codebase for parallel attention.  However, it's not really useful for TMM, so I removed it for this PR. (hence the 'but available').  Let me expand that comment to note it's in the main codebase.

File: torchmultimodal/modules/layers/parallel_blocks.py (hunk 3 above)
Comment (lessw2020) [reply torchmultimodal/modules/layers/parallel_blocks.py:39]: Easy (but I will expand this in the comments to clarify):
num_heads_group_query_attn = 1 and you now have MQA.
num_heads_group_query_attn > 1 and < Q_heads = GQA.  
I should probably also clarify that MQA heads needs to be a multiple of Q heads (there is an assert check so you ultimately can't miss it, but might be nicer to note in docstring).

File: torchmultimodal/modules/layers/parallel_blocks.py (hunk 4 above)
Comment (lessw2020) [reply torchmultimodal/modules/layers/parallel_blocks.py:64]: yes fixed for pretraining. However,  you can do rotary embedding interpolation and expand it though with finetuning (apparently up to 32x).  
https://arxiv.org/pdf/2306.15595.pdf

File: torchmultimodal/modules/layers/position_embedding.py [hunk 5]
```diff
@@ -169,3 +169,98 @@ def forward(self, t: Tensor) -> Tensor:
...
+        Args
+        ----
+        dim : int
+            Per-head embedding dimension
+        max_position_embeddings : int
+            Maximum expected sequence length for the model, if exceeded the cached freqs will be recomputed
+        ratio: int
+            The ratio for the geometric progression to compute the rotation angles
+        """
+        super(RotaryEmbedding, self).__init__()
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchmultimodal/modules/layers/position_embedding.py:192]: good catch, fixed! (though breaking this whole PR into 3 different ones, so fixed in Rotary PR that's coming).

File: torchmultimodal/modules/layers/parallel_blocks.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        self.attention_dropout = nn.Dropout(attention_dropout)
+        self.mlp_dropout = nn.Dropout(projection_dropout)
+
+        # weight init
+        self.num_layers = num_layers
+        self.use_weight_init = use_weight_init
+        if self.use_weight_init:
+            assert (
+                self.num_layers > 1
+            ), f"Need to pass in global num layers for weight init"
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchmultimodal/modules/layers/parallel_blocks.py:108]: it's a valid point - the design was trying to be as portable/encapsulated as possible initially, but with this upstream can perhaps provide a container class to go with it and remove this.

[the remark this answers is not in the document]
Comment (lessw2020) [reply torchmultimodal/modules/layers/parallel_blocks.py:108]: update - I started a wrappers.py class but atm Rohan does not feel we should add it.  Thus, will leave this out for now though will continue to build the wrappers.py class separately (as it has LLMWrapper, ViTWrapper, Encoder, Decoder).
